### ISSUE TITLE
Error Matching and Presentation

### ISSUE THREAD

**Issue by DavidAntaramian** [jonasschmidt/ex_json_schema#17]
I'm interested to hear the background on validation error presentation because I'm currently working to improve the way my company's application handles invalid data matched against schemas. `ex_json_schema` currently returns validations misses in an error tuple with the following spec:

```elixir
{:error, [{String.t, String.t | [String.t | integer]}]}
```

The problem is that the list of errors returned is difficult to act upon since it uses binaries to present the reason. That is, the reason, which is at position 0 in the tuple, isn't guaranteed to have a stable reference. A minimum value violation may have a reason of "Expected the value to be > 4",  "Expected the value to be > 5", etc. That is to say, the value will not be the same and therefore can't be matched upon and acted upon intuitively.

In my mind, it would be more useful to have the validation errors be presented using atoms. For example:

```
{:error,
  [
    {:minimum_violation, [minimum: 4], "#/foo", "Expected the value to be > 4"},
    {:exclusive_minimum_violation, [exclusive_minimum: 5], "#/bar", "Expected the value to be >= 5"},
    {:maximum_length_violation, [maximum_length: 25], "#/person/firstName", "Expected value to have a maximum length of 25 but was 38"}
  ]
}
```

By presenting validation errors in this fashion, it's easy to have my code match on the atom and then take appropriate action based on that. This includes building custom error strings and creating appropriate exceptions.

(Also, for the same reason, I could see making the reference to the path as something like `[:root, "person", "firstName"]`), but that's another matter.

Happy to help out with this, but I wanted to see what others opinions of this was.

**Comment by jonasschmidt** [comment]
Yes, absolutely. This is basically a duplicate of https://github.com/jonasschmidt/ex_json_schema/issues/8 (at least the motivation behind it is the same).

I already have a branch for this (https://github.com/jonasschmidt/ex_json_schema/tree/validation-errors) where the error responses would look like this:

```elixir
{:error, [%ExJsonSchema.Validator.Error{error: %ExJsonSchema.Validator.Error.TypeMismatch{actual: "String", expected: ["Integer"]}, path: "#"}]}
```

The `#/foo/2/bar` JSON Pointer notation is what JSON Schema uses for all references, and it's an official standard with an RFC and all (https://tools.ietf.org/html/rfc6901). There are libraries for working with those as well (https://github.com/xavier/json_pointer and https://github.com/odogono/elixir-jsonpointer for Elixir). So I'm not sure if it's a good idea to give out the path in a different form. What do you think?

I have to iron out the last details in the implementation and decide how the interface would look like to cause minimal breakage for existing code. I had very little time to work on it for the last few months, but there's a chance I will get to finish it next week.

**Comment by DavidAntaramian** [comment]
Ah, great!

My point with JSON pointers is that…well, I have to include yet another library to then process an Elixir data structure (map) against a non-Elixir reference pattern (JSON pointer). By having the pointer destructured into a natively traversable format, I reduce the additional surface area of importing yet another dependency into the application.

**Comment by jonasschmidt** [comment]
I would agree with you if there was a format that makes it natively traversable. There's `get_in` of course, but that can't deal with arrays. So this does not work:

```elixir
get_in(%{"foo" => [%{"bar" => "baz"}]}, ["foo", 0, "bar"])
```

If you know of another way, please let me know.

**Comment by jonasschmidt** [comment]
Improved error presentation and formatting is released in `0.6.0`. The README has some examples of how to use it. I will close this issue for now.
